src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,526 @@
+                else
+                {
+                    // If there's no reference to types declared in this document,
+                    // we will use root node as import container.
+                    root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    containers = ImmutableArray.Create(root);
+                }
+
+                Debug.Assert(containers.Length > 0);
+
+                // Need to import all containing namespaces of old namespace and add them to the document (if it's not global namespace)
+                var namesToImport = CreateAllContainingNamespaces(OldNamespaceParts);
+
+                var optionSet = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+                var placeSystemNamespaceFirst = optionSet.GetOption(GenerationOptions.PlaceSystemNamespaceFirst, document.Project.Language);
+                document = await AddImportsInContainersAsync(
+                    document, 
+                    addImportService, 
+                    containers, 
+                    namesToImport, 
+                    placeSystemNamespaceFirst, 
+                    cancellationToken).ConfigureAwait(false);
+
+                root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                root = _service.ChangeNamespaceDeclaration(root, OldNamespaceParts, NewNamespaceParts);
+
+                // Need to invoke formatter explicitly since we are doing the diff merge ourselves.
+                root = await Formatter.FormatAsync(root, Formatter.Annotation, document.Project.Solution.Workspace, optionSet, cancellationToken)
+                    .ConfigureAwait(false);
+
+                root = root.WithAdditionalAnnotations(Simplifier.Annotation);
+                document = document.WithSyntaxRoot(root);
+                return await Simplifier.ReduceAsync(document, optionSet, cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<Document> FixReferencingDocumentAsync(
+                Document document,
+                IEnumerable<ReferenceLocation> refLocations,    
+                CancellationToken cancellationToken)
+            {
+                // 1. Fully qualify all simple references (i.e. not via an alias) with new namespace.
+                // 2. Add using of new namespace (for each reference's container).
+                // 3. Try to simplify qualified names introduced from step(1).
+
+                var addImportService = document.GetLanguageService<IAddImportsService>();
+                var syncNamespaceService = document.GetLanguageService<ISyncNamespaceService>();
+
+                ImmutableArray<SyntaxNode> containers;
+                (document, containers) = 
+                    await FixReferencesAsync(document, addImportService, syncNamespaceService, refLocations, NewNamespaceParts, cancellationToken)
+                    .ConfigureAwait(false);
```

Comment:
overwriting of 'document' is confusing.  indentation of call chained method is off.